feat: proxy all types of delegated routing

hacdias opened the pull request:
This is my initial version of the "everything bagel" for `someguy`: proxy everything we get from other sources and spit it back when asked.  I did not use that much code from `go-libp2p-routing-helpers` however. Our Delegated Routing stuff uses iterators, while libp2p relies on channels. Makes it very hard to re-use most of the code.

## To Do

- [x] ~Local storage for local caching purposes~ (Extracted to https://github.com/ipfs-shipyard/someguy/issues/25)
- [x] Figure out why the network on my laptop stops working properly if I have this turned on for a few minutes.
    - Only affects laptop, other devices connected to same network are fine.
    - Doesn't seem to happen in all networks.
    - Conclusion: maybe home router has some per-device limit? **Likely**. Disabling accelerated DHT works: https://github.com/ipfs-shipyard/someguy/pull/23#pullrequestreview-1755855029

cc @aschmahmann

aschmahmann reviewed (changes requested):
Thanks for starting this 🙏.

>  Figure out why the network on my laptop stops working properly if I have this turned on for a few minutes.

My guess is running the accelerated dht client isn't making your router happy, but it has some QoS protections (like per-device limits) to protect your devices from each other. Adding a flag to enable/disable the accelerated client seems reasonable to me.

hacdias commented on server_routers.go:
```diff
@@ -0,0 +1,349 @@
+
+func (mi *manyIter[T]) Next() bool {
+	for i, it := range mi.it {
+		if it.Next() {
+			mi.next = i
+			return true
+		}
```
It will, I will take a look. There goes the simplicity 😢

hacdias replied:
Okay, I did something - I did not use the code from routing-helpers as I think it can be done simpler. Please take a look :D

hacdias commented on server_routers.go:
```diff
@@ -0,0 +1,349 @@
+
+type parallelRouter struct {
+	routers []server.ContentRouter
+}
```
I still have to add more tests. It's not as easy to port the tests from libp2p as I thought

hacdias commented:
> My guess is running the accelerated dht client isn't making your router happy, but it has some QoS protections (like per-device limits) to protect your devices from each other. You can use a cli flag to enable/disable the accelerated DHT client and you should see if that makes the difference for you.

 I will check this once at home. I haven't had this issue on the coworking space.

aschmahmann commented on server_routers.go:
```diff
@@ -0,0 +1,383 @@
+	for i, it := range its {
+		go func(ch chan int, it iter.ResultIter[T], index int) {
+			for it.Next() {
+				ch <- index
```
You probably want a `select` that handles `ctx.Done` as well so that if the user doesn't call `Next` but the context cancels you don't get abandoned goroutines.

hacdias commented on server_routers.go:
```diff
@@ -0,0 +1,383 @@
+		it, itErr := call(ri)
+
+		if itErr != nil {
+			err = errors.Join(err, itErr)
```
Personally, I would always prefer that the caller handles the errors. Unfortunately, Boxo's server implementation ignores closing errors:
- https://github.com/ipfs/boxo/blob/08959f281f866bd4197b40bc90cf75c2828c07f0/routing/http/server/server.go#L351
- https://github.com/ipfs/boxo/blob/08959f281f866bd4197b40bc90cf75c2828c07f0/routing/http/server/server.go#L495

So we'd probably have to add it here if we want them.

hacdias commented:
@aschmahmann I've added some tests. They're a bit more basic than the ones we have in https://github.com/libp2p/go-libp2p-routing-helpers, but they should cover enough for now. I think I addressed all the feedback.

I extracted the local storage to a separate issue to tackle in a different PR.

hacdias commented:
Ping @aschmahmann

hacdias commented:
new ping for @aschmahmann ;)

aschmahmann reviewed (changes requested):
Apologies for the delay here @hacdias and thanks for the pings + callouts. Made a few relatively small comments but should be good to go otherwise.

If after the changes you're blocked for me on review just ask @jorropo to review. I'd expect the changes here should be pretty small an easy for me to review though so he doesn't have to load context into the whole PR.

Jorropo commented:
I'm missing something but it seems this code is duplicated from `go-libp2p-routing-helpers`, this is tricky code (I added a fair amount of deadlocks in `go-libp2p-routing-helpers`) I would rather squash bugs once there.

hacdias commented:
> I'm missing something but it seems this code is duplicated

It isn't. It can't be used directly. I had to make many changes. We've discussed this before. I'm merging this as per @aschmahmann approval. We can iterate later.
